src/rust/engine/src/externs/mod.rs

```diff
@@ -332,6 +332,54 @@ pub fn call_method(value: &PyObject, method: &str, args: &[Value]) -> Result<PyO
   value.call_method(gil.python(), method, args_tuple, None)
 }
 
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum InvocationResult {
+  SuccessfulReturn(Value),
+  Exception(Failure),
+}
+
+impl From<Result<PyObject, PyErr>> for InvocationResult {
+  fn from(other: Result<PyObject, PyErr>) -> Self {
+    match other {
+      Ok(obj) => InvocationResult::SuccessfulReturn(Value::from(obj)),
+      Err(err) => InvocationResult::Exception(Failure::from_py_err(err)),
+    }
+  }
+}
+
+impl InvocationResult {
+  pub fn unsafe_extract(self) -> Result<Value, Failure> {
+    match self {
+      InvocationResult::SuccessfulReturn(val) => Ok(val),
+      InvocationResult::Exception(err) => Err(err),
+    }
+  }
+}
+
+impl AsRef<Value> for InvocationResult {
+  fn as_ref(&self) -> &Value {
+    match &self {
+      &InvocationResult::SuccessfulReturn(val) => val,
+      &InvocationResult::Exception(err) => err.as_py_err().unwrap(),
+    }
+  }
+}
+
+impl TryInto<Value> for InvocationResult {
+  type Error = Failure;
+
+  fn try_into(self) -> Result<Value, Failure> {
+    match self {
+      InvocationResult::SuccessfulReturn(val) => Ok(val),
+      InvocationResult::Exception(failure) => failure.into_py_err(),
+    }
+  }
+}
+
+pub fn call_catching_error<T: AsRef<PyObject>>(func: T, args: &[Value]) -> InvocationResult {
```

Comment:
I've generally found it more straightforward to use `&PyObject` as the type of a function parameter, and calling `.as_ref()` on any `Value`s that might need to get passed in.

Reply:
Done!